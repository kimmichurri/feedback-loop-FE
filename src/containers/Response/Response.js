import React, { Component } from 'react'
import ResponseCard from '../ResponseCard/ResponseCard'
export class Response extends Component {
  constructor(props) {
    super(props)
    this.state={
      responses: [],
      displayQuestions: false
    }
  }

  completeSurvey = () => {
    this.setState({
      displayQuestions: !this.state.displayQuestions
    })
  }

  renderQuestions = () => {
    return this.props.questions.map(question => {
      return <ResponseCard key={question.id} question={question} member={this.props.member} currStudent={this.props.currStudent} checkResponse={this.checkResponse} />
    })
  }

  addResponse = (newResponse) => {
    this.setState({
      responses: [...this.state.responses, newResponse]
    })
  }

  updateResponse = (newResponse) => {
    const stateResponses = this.state.responses
    const updatedResponses = stateResponses.map(stateResponse => {
      if(stateResponse.question === newResponse.question) {
        stateResponse = newResponse
      }
      return stateResponse
    })
    this.setState({
      responses: updatedResponses
    })
  }

  checkResponse = (newResponse) => {
    const newResponseId = newResponse.question
    if (this.state.responses.length >= 1) {
      const questionIds = this.state.responses.map(stateResponse => {
        return stateResponse.question
      })
      if(!questionIds.includes(newResponseId)) {
        this.addResponse(newResponse)
      } else {
        this.updateResponse(newResponse)
      }
    } else {
      this.setState({
        responses: [newResponse]
      })
    }
  }

  render() {
    return (
      <div className='member-survey'>
        <button onClick={this.completeSurvey} className=
        'response-button'>
          Give {this.props.member.name} Feedback
        </button>
        {this.state.displayQuestions && this.renderQuestions()}
        <div className='close-button-container'>
          {this.state.displayQuestions && 
            <button 
              onClick={this.completeSurvey} className='close-button'>
              Close Survey for {this.props.member.name}</button>}
        </div>
      </div>
    )
  }
}

export default Response