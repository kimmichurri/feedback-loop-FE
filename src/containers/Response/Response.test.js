import React from 'react'
import { Response } from './Response'
import { shallow } from 'enzyme'

describe('Response', () => {
  let wrapper
  let mockFn = jest.fn()
  let mockMember = { id: 18 }
  let mockQuestions = [
    {
      id: 1,
      questionTitle: "Question One",
      options: [{
          id: "a",
          description: "Option One"
        },
        {
          id: "b",
          description: "Option Two"
        }
      ]
    },
    {
      id: 2,
      questionTitle: "Question Two",
      options: [
        {
          id: "c",
          description: "Option Three"
        },
        {
          id: "d",
          description: "Option Four"
        }
      ]
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<Response 
      member={mockMember}
      questions={mockQuestions}
      collectResponses={mockFn}
    />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  it('should have a default state', () => {
    const expected = {
      responses: [],
      displayQuestions: false,
      saveResponses: false,
    }
    expect(wrapper.state()).toEqual(expected)
  })

  it('should update state when completeSurvey is called', () => {
    wrapper.instance().completeSurvey()
    expect(wrapper.state('displayQuestions')).toBe(true)
  })

  it('should return ResponseCard components for each question in props', () => {
    const result = wrapper.instance().renderQuestions()
    expect(result).toHaveLength(2)
  })

  it('should set state with a new response if there are no responses in state', () => {
    wrapper.setState({
      responses: []
    })

    const newResponse = {
      question: 1,
      answer: 2,
      member: 4
    }
    wrapper.instance().checkResponse(newResponse)
    expect(wrapper.state('responses')).toEqual([newResponse])
  })

  it('should call addResponse if there are responses in state', () => {
    const addResponseSpy = jest.spyOn(wrapper.instance(), 'addResponse')
    wrapper.setState({
      responses: [{
        question: 15,
        answer: 2,
        member: 4
      }]
    })
    const newResponse = {
      question: 1,
      answer: 2,
      member: 4
    }
    wrapper.instance().checkResponse(newResponse)
    expect(addResponseSpy).toHaveBeenCalledWith(newResponse)
  })

  it('should call updateResponse if there are responses in state', () => {
    const updateResponseSpy = jest.spyOn(wrapper.instance(), 'updateResponse')
    wrapper.setState({
      responses: [{
        question: 15,
        answer: 2,
        member: 4
      }]
    })
    const newResponse = {
      question: 15,
      answer: 2,
      member: 4
    }
    wrapper.instance().checkResponse(newResponse)
    expect(updateResponseSpy).toHaveBeenCalledWith(newResponse)
  })

  it('should add a response to state when addResponse is called', () => {
    wrapper.setState({
      responses: [{
        question: 15,
        answer: 2,
        member: 4
      }]
    })
    const newResponse = {
      question: 1,
      answer: 2,
      member: 4
    }
    const expected = [
      {
        question: 15,
        answer: 2,
        member: 4
      },
      {
        question: 1,
        answer: 2,
        member: 4
      }
    ]
    wrapper.instance().addResponse(newResponse)
    expect(wrapper.state('responses')).toEqual(expected)
  })

  it('should update the correct response in state when updateResponse is called', () => {
    wrapper.setState({
      responses: [{
        question: 15,
        answer: 2,
        member: 4
      }]
    })
    const newResponse = {
      question: 15,
      answer: 3,
      member: 4
    }
    const expected = [
      {
        question: 15,
        answer: 3,
        member: 4
      }
    ]
    wrapper.instance().updateResponse(newResponse)
    expect(wrapper.state('responses')).toEqual(expected)
  })

  it('should call warnToast when closeResponse is called', () => {
    const warnToastSpy = jest.spyOn(wrapper.instance(), 'warnToast')
    wrapper.setState({
      responses: [ { question: 4 } ]
    })
    wrapper.instance().closeResponse()
    expect(warnToastSpy).toHaveBeenCalled()

  })

  it('should toggle displayQuestions when closeResponse is called', () => {
    wrapper.setState({
      responses: [ { question: 1 }, { question: 2 } ],
      displayQuestions: true
    })
    wrapper.instance().closeResponse()
    expect(wrapper.state('displayQuestions')).toBe(false)
  })
})