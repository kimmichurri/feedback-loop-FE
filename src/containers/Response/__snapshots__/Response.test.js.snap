// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response should match the snapshot 1`] = `
<div
  className="member-survey"
>
  <button
    className="response-button"
    onClick={[Function]}
  >
    Give 
     Feedback
  </button>
  <div
    className="close-button-container"
  />
</div>
`;
